### PR TITLE
first sketch for callable factorisations

### PR DESCRIPTION
**Pull request by fchapoton**
Implement a simple substitution mechanism for "Factorization" instances, used when factorising polynomials for example.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/structure/factorization.py [hunk 1]
```diff
@@ -567,11 +568,11 @@ def base_change(self, U):
...
             return Factorization([(U(f[0]), f[1]) for f in list(self)], unit=U(self.unit()))
         except TypeError:
-            raise TypeError("Impossible to coerce the factors of %s into %s"%(self, U))
+            raise TypeError("Impossible to coerce the factors of %s into %s" % (self, U))
 
-    def is_commutative(self):
+    def is_commutative(self) -> bool:
         """
-        Return True if my factors commute.
+        Return ``True`` if my factors commute.
```
Comment (dcoudert) [inline src/sage/structure/factorization.py]: In many places, we have instead: `Check whether the factors commute.`

File: src/sage/structure/factorization.py [hunk 2]
```diff
@@ -219,24 +219,25 @@ def __init__(self, x, unit=None, cr=False, sort=True, simplify=True):
 
         INPUT:
 
-        - ``x`` - a list of pairs (p, e) with e an integer;
-          otherwise a TypeError is raised
+        - ``x`` -- a list of pairs (p, e) with e an integer;
+          otherwise a :class:`TypeError` is raised
 
-        - ``unit`` - (default: 1) the unit part of the factorization.
+        - ``unit`` -- (default: 1) the unit part of the factorization.
```
Comment (dcoudert) [inline src/sage/structure/factorization.py]: missing `;` after parenthesis (many places).
```diff
- ``unit`` -- (default: 1) the unit part of the factorization.
+ ``unit`` -- (default: 1); the unit part of the factorization.
```

File: src/sage/structure/factorization.py [hunk 3]
```diff
@@ -1270,21 +1328,21 @@ def lcm(self, other):
...
+                s[a] = max(d1.get(a, 0), d2.get(a, 0))
             return Factorization(list(s.items()))
         else:
             raise NotImplementedError("lcm is not implemented for non-commutative factorizations")
 
-    def is_integral(self):
+    def is_integral(self) -> bool:
         r"""
-        Return True iff all exponents of this Factorization are non-negative.
+        Return whether all exponents of this Factorization are non-negative.
```
Comment (dcoudert) [inline src/sage/structure/factorization.py:1345]: Could be
```diff
- Return whether all exponents of this Factorization are non-negative.
+ Check whether all exponents of this Factorization are non-negative.
```

**Review by dcoudert — CHANGES_REQUESTED** [review]

**Review by dcoudert — APPROVED** [review]
LGTM.

**Comment by fchapoton** [comment]
Hmm, an interesting new bug has followed:
```
sage: g=(q-1)**-1
sage: factor(g)
(q - 1)^-1
sage: _(q=1)
0
```

**Pull request merged**